### PR TITLE
5568 show customer for program by tag

### LINKED ISSUE #5431: When the supplier and the customer have different master list for the same program, you need to display the customer one when creating a manual requisition
## What went wrong? 😲

In Ivory Coast context, the same program have more than one master list. For the same program, there is one master list for the District, one for the Health centers (CS) and one for the Hospitals. They share a lot of items but there are some differences between each master list as you can see below there are items present in the Health center master list (in red) that are not in the District master list.
 
https://drive.google.com/file/d/1pPJYRrta-IGA5h4yxrCUqFbErBrP2GwC/view?usp=sharing

So Normally, you do a request requisition (internal order) using your master list ie the Health Center (CS) will use its master list to create an internal order to send to the District. So here you can see that all the requisitions have the master list version of the program that belongs to the Health Center that did the internal order.
 
https://drive.google.com/file/d/1Dnw6wLr-GDYCodFxc9qCZ-Hv2Aij1z1B/view?usp=sharing

But when you do a manual requisition for a customer you need to use the program master list of the customer rather than your own and that is why in mSupply OG, you choose the customer first and then the program. See the video below:

https://drive.google.com/file/d/1gb2oxkwDlRLptk3tFjrCaW8MyK-4BAQ8/view?usp=sharing

Currently in omSupply when you do a manual requisition you have to choose the program master list first and my only choice is the master list of the district but I should see the master list of the Health Center (CS)
https://drive.google.com/file/d/1kitLmNbEAzwzT7EHpVxg6CPJNXxEyhET/view?usp=sharing

The second problem, this causes is that by using the Supplier master list instead of the customer master list is that the list of customers is populated by other stores/facilities that uses that master list to order from a higher or equal level facility (See below)
https://drive.google.com/file/d/1n34Y86cIiJzR_LDbqcWnWeLzj6neSSom/view?usp=sharing

## Expected behaviour 🤔
You should be able to choose the master list that belongs to the customer instead of the master list that belongs to the supplier and for this you need to choose the customer before you choose the program.

## How to Reproduce 🔨

Steps to reproduce the behaviour:

1. Go to Requisitions
2. Click on Create new requisition
3. Choose program 
4. See error. The supplier master list is showing instead of the customer one

## Your environment 🌱

- Open mSupply Version: 2.4.0
- Legacy mSupply Central Server Version: v19-0dev27

- Platform: Android 14

- Database type: SqLite 3

### LINKED ISSUE #5568: Program customer selection don't take name tag into consideration and will show if customer is using master list
## What went wrong? 😲
Program customer selection don't take name tag into consideration and will show if customer is using master list 

## How to Reproduce 🔨

Steps to reproduce the behaviour:

1. Create a program master list 
2. Have a customer use the master list but don't assign the tag to customer
3. Go to OMS and click Requisition
4. Try add new program requisition
5. See that you can select customer even if they don't have tag

## Your environment 🌱

- Open mSupply Version: 2.4.0
- Legacy mSupply Central Server Version:

- Platform:

- Database type:

### PR DESCRIPTION
**Pull request by roxy-dao**
Fixes #5568

# 👩🏻‍💻 What does this PR do?
Ignore branch number... 

Allow customer selection for program based on name tags.

https://github.com/user-attachments/assets/cab61aa9-73a3-43ad-8275-7e098fccc034

## 💌 Any notes for the reviewer?

# 🧪 Testing

- [ ] Have multiple customers
- [ ] Set up some program master lists
- [ ] Have customers use program master list
- [ ] Add the tags to the customers (make sure to have some customers use a program master list without the associated tag)
- [ ] Go to OMS requisition
- [ ] Click `New Requisition`
- [ ] Customer selection for the program master lists should match with customers who you have assigned the name tags to

# 📃 Documentation

- [x] **Part of an epic**: documentation will be completed for the feature as a whole
- [ ] **No documentation required**: no user facing changes or a bug fix which isn't a change in behaviour
- [ ] **These areas should be updated or checked**: 
  1.
  2.

### CODE REVIEW

File: server/repository/src/db_diesel/name_tag.rs [hunk 1]
```diff
@@ -2,11 +2,8 @@ use diesel::prelude::*;
```
Comment (roxy-dao) [inline server/repository/src/db_diesel/name_tag.rs:2]: Thought I'd refactor this to not use dsl alias while I was at it...

File: server/service/src/requisition/program_settings/customer_program_settings/prepare.rs [hunk 2]
```diff
@@ -63,15 +64,19 @@ pub(super) fn prepare_customer_program_settings(
...
 
-    // Customers, which are visible in current store and have these programs (this is determined by having program master list visible)
     let filter = ProgramCustomerFilter::new().program_id(EqualFilter::equal_any(program_ids));
-    let program_customers =
+    let mut program_customers =
         ProgramCustomerRepository::new(&ctx.connection).query(store_id, filter)?;
+    program_customers.sort_by(|a, b| a.program.id.cmp(&b.program.id));
+    program_customers.dedup_by(|a, b| {
+        a.program.id == b.program.id && a.customer.name_row.id == b.customer.name_row.id
+    });
```
Comment (roxy-dao) [inline server/service/src/requisition/program_settings/customer_program_settings/prepare.rs:79]: Removing duplicate program customers

File: server/service/src/requisition/response_requisition/update.rs [hunk 3]
```diff
@@ -283,23 +283,6 @@ mod test_update {
             ),
             Err(ServiceError::CannotEditRequisition)
         );
-
-        // ReasonRequired when requested differs from suggested
-        if let Err(ServiceError::ReasonsNotProvided(errors)) = service.update_response_requisition(
```
Comment (roxy-dao) [inline server/service/src/requisition/response_requisition/update.rs:288]: TODO: Add test back in. This one fails

**Review by jmbrunskill — APPROVED** [review]
I'm not a big fan of having to de-dup the query in rust, although it's probably a big problem I guess. 

I suspect if you call distinct() on your query the db will do it for you, but i haven't tested properly.
image

For situations like this where you need to filter on some additional data but not actually return it, it's probably better to do a sub query rather than joins.

I have a go at doing it with Diesel but ran into some problems. Don't think it's worth fighting it further, you may have had similar issues?

Keen to get that test fixed though if it's related to this change? If it's unrelated can just add it back in then raise a separate issue to fix?

I haven't tested properly but I'm thinking QA can do that?

File: server/service/src/requisition/response_requisition/update.rs (hunk 3 above)
Comment (roxy-dao) [reply server/service/src/requisition/response_requisition/update.rs:288]: Probably, but will add TODO to fix later

File: server/repository/src/db_diesel/diesel_schema.rs [hunk 4]
```diff
@@ -38,6 +58,7 @@ allow_tables_to_appear_in_same_query!(
     program_requisition_settings,
     program_requisition_order_type,
     period,
+    name_tag,
```
[the remark this answers is not in the document]
Comment (roxy-dao) [reply server/repository/src/db_diesel/diesel_schema.rs:61]: The name table joins a lot of other tables, so I just added it in there to get all the joins >.>.
Sorry, we should really refactor this file!

File: server/service/src/requisition/response_requisition/update.rs [hunk 5]
```diff
@@ -112,11 +112,11 @@ pub fn validate(
         )),
     )?;
 
-    let prefs = get_store_preferences(connection, &store_id)?;
+    let prefs = get_store_preferences(connection, store_id)?;
 
     if requisition_row.program_id.is_some()
         && prefs.extra_fields_in_requisition
-        && reason_options.len() > 0
+        && !reason_options.is_empty()
```
[the remark this answers is not in the document]
Comment (roxy-dao) [reply server/service/src/requisition/response_requisition/update.rs:119]: I like it like this 🤣  might just be me

**Pull request merged**